demos/common/cpp_gapi/utils_gapi/include/utils_gapi/onnx_backend.hpp

```diff
@@ -0,0 +1,120 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <functional>
+#include <list>
+#include <map>
+#include <stdexcept>
+#include <string>
+
+#include "utils_gapi/backend_description.hpp"
+
+#ifdef GAPI_ONNX_BACKEND
+#include <opencv2/gapi/infer/onnx.hpp>
+
+const std::initializer_list<std::string> getONNXSupportedEP();
+
+template<class ExecNetwork, class ...Args>
+void applyONNXProviders(ExecNetwork& net, const ModelConfig &config, inference_backends_t backend_cfgs = inference_backends_t{});
+
+template<class ExecNetwork>
+struct BackendApplicator<ExecNetwork,
+                         cv::gapi::onnx::Params> {
+    static cv::gapi::GNetPackage apply(const std::string &model_path, const ModelConfig & config, const inference_backends_t &backends) {
+        auto net =
+            cv::gapi::onnx::Params<ExecNetwork>{
+                model_path
+            };
+        applyONNXProviders(net, config, backends);
+        return cv::gapi::networks(net);
+    }
+};
+
+namespace {
+struct CPUProvider{};
+template<class ExecNetwork, class Provider = CPUProvider>
+struct ProviderApplicator {
+    static void apply(ExecNetwork &, const ModelConfig &, const inference_backends_t &backends) {
+        static_assert(std::is_same<Provider, CPUProvider>::value, "Unsupported ONNX provider requested. Please add a partial specialization if required");
+        if (backends.size() > 1) {
+            throw std::runtime_error("ONNX CPU execution provider must be only provider in the provider list or "
```

Comment:
`CPU` provider isn't even exposed to `onnx::Params` I'd not mentioned it there at all

Reply:
If user defines `OVEP,DML` order it means that everything that don't map on these two will be transferred on `CPU` automatically. Despite user can't specify `OVEP,CPU,DML` since `CPU` supports everything there will be nothing to offload on `DML` in this case.

Reply:
`--backend ONNX/CPU` was added behind `--backend ONNX` by reason of consistency to align with `ONNX/DMl`, `ONNX-OVEP`. When I have typed `ONNX/CPU`  and it would have throw me an error that `CPU provider is not supported` - then I wonder what provider is implied by `ONNX`

> . Despite user can't specify OVEP,CPU,DML since CPU supports everything there will be nothing to offload on DML in this case.

The condition above checks such scenario and throws error, that such combinations do not make sense. 
The only one stuff is seemed to be fixed here is removing ` or be the last in order`
@TolyaTalamanov  do you agree?

Reply:
1. `--backend ONNX` - Is this valid? Will be used default `CPU` backend?
2. `--backend ONNX/CPU` - I assume just `CPU` EP
3. `--backend ONNX/CPU/CPU` - is that valid? (I hope no)
4. `--backend ONNX/OV/CPU` - how would you figure out if `CPU` is the device for `OV` but not the second execution provider?
5. `--backend ONNX/OV/CPU/DML` - what is that line mean? I hope it will throw since `CPU` can be in the middle if it's `EP` (but look at the question in `4`)

Reply:
> --backend ONNX - Is this valid? Will be used default CPU backend? 

ONNX on CPU is used

> --backend ONNX/CPU - I assume just CPU EP

ONNX on CPU is used

> --backend ONNX/CPU/CPU - is that valid? (I hope no)

Please check on code: https://github.com/openvinotoolkit/open_model_zoo/pull/3853/files#diff-a56ed8974441e654093998a6da1c245515cc44c781788d9b69061efeeb9a4692R49
any additional number of arguments is forbidden for ONNX CPU provider

> --backend ONNX/OV/CPU - how would you figure out if CPU is the device for OV but not the second execution provider?

From code:
List of backends  are parsed using `,` separator
https://github.com/openvinotoolkit/open_model_zoo/pull/3853/files#diff-b78e9f5bf604e1f15cbaea07f569ff7e9fd050fa3fc1ea6beb64c2333d840e90R66

List of attributes belong to specific backend are parsed using `/` https://github.com/openvinotoolkit/open_model_zoo/pull/3853/files#diff-714e493f5d95803493777c17cde54eb9637fdbef2354af62c990fc86ef1be889R21

So, ONNX- is backend,  OV - is onnx provider, CPU - is a provide args meant device
> --backend ONNX/OV/CPU/DML

is impossible here